### PR TITLE
How to implement Caching in AdonisJS 5

### LINKED ISSUE #2001: The Best Caching Strategy for AdonisJS RESTFUL API
### Proposed title of article
A Caching Strategy for AdonisJS RESTFUL API

### Introduction paragraph (2-3 paragraphs):
I will be showing you the best practice in handling Caching in Adonisjs, discussing in details the best strategies involved in handling a scalable Caching System in your RESTFUL API. We will build a real-world app to demonstrate how to boost performance with a good Caching Strategy.

### Key takeaways: 
Students will learn how to build a high-performance app with Adonisjs
Students will learn how to implement a Caching System in AdonisJS Restful API
Students will learn how to choose the right Caching Strategy depending on the project.
Students will learn about the concept of Caching and how to scale large App with it.
Students will learn best practices in building performance-oriented Web apps with AdonisJS.

### References: 
https://dev.to/vigzmv/a-simple-caching-strategy-for-node-rest-apis-part-1-72a
https://blog.appsignal.com/2021/03/10/powerful-caching-with-redis-for-nodejs-applications.html
https://hackernoon.com/how-to-setup-caching-in-nodejs-using-redis-tq143usy

### PR DESCRIPTION
**Pull request by Kaperskyguru**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

This fixes #2001 

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by Kaperskyguru** [comment]
@geoffrey45 Thanks very much for your help in the other PR. I have deleted and created this new PR. Thanks.

**Comment by geoffrey45** [comment]
> Hello @geoffrey45 
> 
> What is the status of this content?
> 
> Thanks.

@Kaperskyguru I'm working on your article.

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
+# How to Implement Caching in AdonisJS 5.
+
+The performance of a web application is becoming an integral part of the application as many statistics have shown that slow web applications can result in loss of traffic, sales, and incomes in many different ways.
+
+In this article, we will learn how to implement a Caching System on our Forum API developed with AdonisJS and integrate the newly created Caching Package for the new AdonisJS 5 with optimum Caching experience.
```
Comment (geoffrey45) [inline content/articles/how-to-implement-caching-in-adonisjs-5/index.md:5]: Use consistent capitalization. eg. The words "Caching" and "System" should be lower case. Only capitalize nouns. Please revise your document.

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+```bash
+node ace invoke @kaperskyguru/adonis-cache
+//Or
+node ace configure @kaperskyguru/adonis-cache
+```
+
+Next, open the `.env` file, inform the Cache Package to use `redis` as our Cache Driver, and move below to configure our Redis server, so our Cache Package can communicate successfully to the Redis Server.
+
+    CACHE_DRIVER=redis
+
```
Comment (geoffrey45) [inline content/articles/how-to-implement-caching-in-adonisjs-5/index.md:62]: Use a code block here

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    await Cache.set('ticket_id_' + ticket.id, ticket, 60)
+    return forum
+}
+```
+
+The Write Operations first goes to the Database (Write Around) before updating the Cache server during the Read Operation or immediately to speed up the Read operations subsequently.
+
+### Step 3 → Testing the Performance
+
+Now that we have implemented our Caching System in our project, let's run some tests and see how the performance of our API is improved.
```
Comment (geoffrey45) [inline content/articles/how-to-implement-caching-in-adonisjs-5/index.md:201]: Mention what you are using to test the performance and how we can get it.

**Review by geoffrey45 — CHANGES_REQUESTED** [review]
@Kaperskyguru | Your article looks great. Please look at the attached comments.

**Comment by Kaperskyguru** [comment]
@geoffrey45 I have submitted a new PR to fix the issues you raised. Thanks for reviewing.

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+We will continue with the Forum API we have already developed with the article above. you can [clone the repository](https://github.com/Kaperskyguru/adonisjs-forum-api) to get started.
+
+### Setting up Adonis Cache
+After installing and setting up your Adonis 5 RestFul API project, we will install and configure our caching package. This will simplify implementing our caching strategy to the Forum API, to improve the performance or response time of the API.
+
+Run the following command to install the Adonis cache package.
+
+```bash
+npm i @kaperskyguru/adonis-cache
+```
```
Comment (geoffrey45) [inline content/articles/how-to-implement-caching-in-adonisjs-5/index.md:45]: Tell us more about this package: `@kaperskyguru/adonis-cache`.

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+In the world of caching, there are many caching strategies and methods available depending on the structure of your data structure and the different operations performed on the data. You can choose other caching strategies depending on the features and type of the project.
+
+You can read through the [best caching Strategy for Adonis 5](https://masteringbackend.com/posts/caching-strategy-for-restful-api/) which will show you the different caching strategies, and help you choose which one to use depending on the features or the type of project.
+
+We will be using the **cache aside strategy** for all the read operations and a mixture of **write Around** and **write through strategy** for the write operations.
+
+Open your Controller file. 
+
+We already have the `TicketsController` file found in our previous project's `app/Controllers/Http` folder.
+
```
Comment (geoffrey45) [inline content/articles/how-to-implement-caching-in-adonisjs-5/index.md:106]: In the specified folder, there's no `TicketsController.ts`. The only file that resembles this is the `ForumsController.ts`.

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+                }
+            } catch (error) {
+                console.log(error)
+            }
+        }
+    }
+```
+
+The read operations are shown mainly in the `index()` and `show` methods for retrieving all forums and a single forum by ID.
+
```
Comment (geoffrey45) [inline content/articles/how-to-implement-caching-in-adonisjs-5/index.md:151]: There is no `show()` method in the code block above.

**Review by geoffrey45 — CHANGES_REQUESTED** [review]
@Kaperskyguru I'm almost through the review. Please look at the attached comments.

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
+# How to Implement caching in Adonis.js 5.
+The performance of a web application is an integral part of the application. Many statistics have shown that slow web applications can result in loss of traffic, sales, and incomes in many different ways.
+
+In this article, we will learn how to implement a caching system on our Forum API developed with Adonis. We will also integrate the newly created caching package for the new Adonis 5 with optimum caching experience.
+
```
[the remark this answers is not in the document]
Comment (Kaperskyguru) [reply content/articles/how-to-implement-caching-in-adonisjs-5/index.md:5]: Done

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+Getting started with Adonis 5 is very easy and straightforward as it is well-documented.
+
+If you're just getting started with Adonis 5, I recommend reading [AdonisJS 5 tutorial: The Ultimate Guide](https://masteringbackend.com/posts/adonisjs-tutorial-the-ultimate-guide) and [building a restful API with Adonis 5](/engineering-education/build-a-restful-api-with-adonisjs/) to get to the speed.
+
+You can run this command to have Adonis 5 installed.
+
+```bash
+npm init adonis-ts-app@latest forum-api
+```
+
```
[the remark this answers is not in the document]
Comment (Kaperskyguru) [reply content/articles/how-to-implement-caching-in-adonisjs-5/index.md:35]: Done

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+We will continue with the Forum API we have already developed with the article above. you can [clone the repository](https://github.com/Kaperskyguru/adonisjs-forum-api) to get started.
+
+### Setting up Adonis Cache
+After installing and setting up your Adonis 5 RestFul API project, we will install and configure our caching package. This will simplify implementing our caching strategy to the Forum API, to improve the performance or response time of the API.
+
```
[the remark this answers is not in the document]
Comment (Kaperskyguru) [reply content/articles/how-to-implement-caching-in-adonisjs-5/index.md:40]: done

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md (hunk 4 above)
Comment (Kaperskyguru) [reply content/articles/how-to-implement-caching-in-adonisjs-5/index.md:45]: Done

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md (hunk 5 above)
Comment (Kaperskyguru) [reply content/articles/how-to-implement-caching-in-adonisjs-5/index.md:106]: Done

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md (hunk 6 above)
Comment (Kaperskyguru) [reply content/articles/how-to-implement-caching-in-adonisjs-5/index.md:151]: DOne

**Comment by Kaperskyguru** [comment]
@geoffrey45 I have just submitted a PR to fix the issues you raised. Thanks for reviewing.

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+            // This code is only executed if there was a MISS
+            return await Forum.query().preload('user').preload('posts');
+
+        })
+        return forums
+    }
+
+    // Reading a Single Forum by ID
+    public async update({ auth, request, params}: HttpContextContract){
+        try {
```
Comment (geoffrey45) [inline content/articles/how-to-implement-caching-in-adonisjs-5/index.md:138]: Unused variable `auth`. Please remove it.

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+                forum.title = request.input('title');
+                forum.description = request.input('description');
+                if (await forum.save()) {
+                    await forum.preload('user')
+                    await forum.preload('posts')
+                    return forum
+                }
+            } catch (error) {
+                console.log(error)
+            }
```
Comment (geoffrey45) [inline content/articles/how-to-implement-caching-in-adonisjs-5/index.md:154]: This catch block does not have a matching try block.

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+
+        })
+        return forums
+    }
+
+    // Reading a Single Forum by ID
+    public async update({ auth, request, params}: HttpContextContract){
+        try {
+            const ticket = await Cache.remember('forum_id_' + params.id, 60, async function () {
+            // This code is only executed if there was a MISS
```
Comment (geoffrey45) [inline content/articles/how-to-implement-caching-in-adonisjs-5/index.md:140]: You have defined `ticket` but referred to `forum` in this code block.

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+// Tries to find a Forum from Cache, if found returns the Forum
+    const forum = await Cache.remember('forum_id_' + params.id, 60, async function () {
+        // If not found, Retrieves from Database and Save to Cache.
+        return await Forum.find(params.id)
+    })
+
+    if (forum && await ticket.save()) {
+
+        // If updates successfully in database, then updates the Cache Server.
+        await Cache.update('forum_id_' + params.id, ticket, 60)
```
Comment (geoffrey45) [inline content/articles/how-to-implement-caching-in-adonisjs-5/index.md:180]: `ticket` is undefined.

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+            await forum.preload('user')
+            await forum.preload('posts')
+            return forum
+        }
+        return // 422
+    }
+}
+
+public async store({ auth, request, response}: HttpContextContract)
+{
```
Comment (geoffrey45) [inline content/articles/how-to-implement-caching-in-adonisjs-5/index.md:194]: Unused variable `response`.

File: content/articles/how-to-implement-caching-in-adonisjs-5/index.md [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+    const user = await auth.authenticate();
+    const forum = new Forum();
+    forum.title = request.input('title');
+    forum.description = request.input('description');
+    await user.related('forums').save(forum)
+
+    // Stores a new Ticket to Cache
+    await Cache.set('ticket_id_' + ticket.id, ticket, 60)
+    return forum
```
Comment (geoffrey45) [inline content/articles/how-to-implement-caching-in-adonisjs-5/index.md:203]: `ticket` is undefined.

**Review by geoffrey45 — APPROVED** [review]
@Kaperskyguru There are some few minor issues with your code. Please see the attached comments.

**Comment by Kaperskyguru** [comment]
Hello @geoffrey45, I just send in a new commit to fix all your issues.

Thanks for reviewing.

**Comment by hectorkambow** [comment]
[How to implement Caching in AdonisJS 5 #2447.pdf](https://github.com/section-engineering-education/engineering-education/files/6697324/How.to.implement.Caching.in.AdonisJS.5.2447.pdf)

👋 Good afternoon - Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.
Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
Be sure to see our resources page to see more info on plagiarism and what is considered as such.

**Comment by Kaperskyguru** [comment]
Hello @geoffrey45 @hectorkambow,

I just pushed in a commit to fix the Plagiarism flag.

Thanks.

**Comment by hectorkambow** [comment]
@geoffrey45  

Please let me know if this article is ready for final review?

**Comment by hectorkambow** [comment]
@Kaperskyguru 
Great work gents 🚀 
@geoffrey45 
👏

**Pull request merged**